src/ring_theory/graded_algebra/basic.lean

```diff
@@ -77,4 +82,98 @@ def graded_algebra.decompose : A ≃ₐ[R] ⨁ i, 𝒜 i := alg_equiv.symm
+  graded_algebra.decompose 𝒜 x = direct_sum.of _ i (⟨x, hx⟩ : 𝒜 i) :=
+graded_algebra.decompose_coe _ ⟨x, hx⟩
+
+lemma graded_algebra.decompose_of_mem_same {x : A} {i : ι} (hx : x ∈ 𝒜 i) :
+  (graded_algebra.decompose 𝒜 x i : A) = x :=
+by rw [graded_algebra.decompose_of_mem _ hx, direct_sum.of_eq_same, subtype.coe_mk]
+
+lemma graded_algebra.decompose_of_mem_ne {x : A} {i j : ι} (hx : x ∈ 𝒜 i) (hij : i ≠ j):
+  (graded_algebra.decompose 𝒜 x j : A) = 0 :=
+by rw [graded_algebra.decompose_of_mem _ hx, direct_sum.of_eq_of_ne _ _ _ _ hij, submodule.coe_zero]
+
+
+variable [Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0)]
+
+lemma graded_algebra.mem_support_iff
+  (r : A) (i : ι) :
+i ∈ graded_algebra.support 𝒜 r ↔ (graded_algebra.proj 𝒜 i r ≠ 0) :=
+begin
+  rw [graded_algebra.support, dfinsupp.mem_support_iff, graded_algebra.proj_apply],
+  simp only [ne.def, submodule.coe_eq_zero],
+end
+
+lemma graded_algebra.sum_support_decompose (r : A) :
+  ∑ i in graded_algebra.support 𝒜 r, (graded_algebra.decompose 𝒜 r i : A) = r :=
+begin
+  conv_rhs {
+    rw [←(graded_algebra.decompose 𝒜).symm_apply_apply r,
+        ←direct_sum.sum_support_of _ (graded_algebra.decompose 𝒜 r)] },
+  rw [alg_equiv.map_sum, graded_algebra.support],
+  simp_rw graded_algebra.decompose_symm_of,
+end
+
+lemma graded_algebra.mul_decompose (r r' : A) (i : ι) :
+  (graded_algebra.decompose 𝒜 (r * r') i : A) =
+  ∑ ij in finset.filter (λ ij : ι × ι, ij.1 + ij.2 = i)
+    ((graded_algebra.support 𝒜 r).product (graded_algebra.support 𝒜 r')),
+    (graded_algebra.decompose 𝒜 r ij.1 : A) * (graded_algebra.decompose 𝒜 r' ij.2 : A) :=
+begin
+  dsimp only [graded_algebra.proj_apply, graded_algebra.support],
+  rw [alg_equiv.map_mul, direct_sum.mul_eq_sum_support_ghas_mul],
+  generalize : (graded_algebra.decompose 𝒜) r = a,
+  generalize : (graded_algebra.decompose 𝒜) r' = a',
+  rw [dfinsupp.finset_sum_apply, submodule.coe_sum],
+
+  have set_eq : (dfinsupp.support a).product (dfinsupp.support a') =
+    finset.filter (λ ij : ι × ι, ij.1 + ij.2 = i) _ ∪
+    finset.filter (λ ij : ι × ι, ij.1 + ij.2 ≠ i) _ := (finset.filter_union_filter_neg_eq _ _).symm,
+  conv_lhs { rw [set_eq] },
+  rw [finset.sum_union],
+  suffices : ∀ a b c : A, b = 0 → a = c → a + b = c, apply this,
+
+  apply finset.sum_eq_zero, rintros ⟨j, k⟩ h,
+  simp only [ne.def, finset.mem_filter, finset.mem_product] at h ⊢,
+  obtain ⟨⟨h₁, h₂⟩, h₃⟩ := h,
+  simp only [submodule.coe_eq_zero],
+  rw direct_sum.of_eq_of_ne, exact h₃,
+
+  rw finset.sum_congr rfl,
+  rintros ⟨j, k⟩ h, simp only [finset.mem_filter, finset.mem_product] at h ⊢,
+  obtain ⟨⟨h₁, h₂⟩, h₃⟩ := h,
+  rw [←h₃, direct_sum.of_eq_same], refl,
+
+  intros x y z h1 h2, rw [h1, h2, add_zero],
+
+  rw [finset.disjoint_iff_inter_eq_empty, finset.eq_empty_iff_forall_not_mem],
+  rintros ⟨j, k⟩ rid,
+  simp only [ne.def, finset.mem_filter, finset.mem_inter, finset.mem_product] at rid,
+  rcases rid with ⟨⟨_, h₁⟩, ⟨_, h₂⟩⟩, exact h₂ h₁,
+end
```

Comment:
With my proposed `coe_of_submodule` lemma:
```suggestion
lemma graded_algebra.mul_decompose (r r' : A) (i : ι) :
  (graded_algebra.decompose 𝒜 (r * r') i : A) =
  ∑ ij in finset.filter (λ ij : ι × ι, ij.1 + ij.2 = i)
    ((graded_algebra.support 𝒜 r).product (graded_algebra.support 𝒜 r')),
    (graded_algebra.decompose 𝒜 r ij.1 : A) * (graded_algebra.decompose 𝒜 r' ij.2 : A) :=
begin
  dsimp only [graded_algebra.proj_apply, graded_algebra.support],
  rw [alg_equiv.map_mul, direct_sum.mul_eq_sum_support_ghas_mul],
  rw [dfinsupp.finset_sum_apply, submodule.coe_sum],
  simp_rw [direct_sum.coe_of_submodule, ←finset.sum_filter, set_like.coe_ghas_mul],
end
```

Reply:
In fact, this is really just a special case of:
```
lemma direct_sum.coe_mul_apply_submodule (r r' : ⨁ i, 𝒜 i) (i : ι) :
  ((r * r') i : A) =
    ∑ ij in finset.filter (λ ij : ι × ι, ij.1 + ij.2 = i) (r.support.product r'.support),
      r ij.1 * r' ij.2 :=
begin
  rw [direct_sum.mul_eq_sum_support_ghas_mul, dfinsupp.finset_sum_apply, submodule.coe_sum],
  simp_rw [direct_sum.coe_of_submodule, ←finset.sum_filter, set_like.coe_ghas_mul],
end
```
which only needs `set_like.graded_monoid` instead of `graded_algebra`.